core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1297,11 +1332,188 @@ public TimeValue getInactiveTime() {
+    }
+
+    /**
+     * Notifies the service of the current allocation ids in the cluster state.
+     * see {@link GlobalCheckpointService#updateAllocationIdsFromMaster(Set, Set)} for details.
+     *
+     * @param activeAllocationIds       the allocation ids of the currently active shard copies
+     * @param initializingAllocationIds the allocation ids of the currently initializing shard copies
+     */
+    public void updateAllocationIdsFromMaster(Set<String> activeAllocationIds, Set<String> initializingAllocationIds) {
+        ensurePrimary();
+        Engine engine = getEngineOrNull();
+        // if engine is not yet started, we are not ready yet and can just ignore this
+        if (engine != null) {
+            engine.seqNoService().updateAllocationIdsFromMaster(activeAllocationIds, initializingAllocationIds);
+        }
+    }
+
+    abstract class ScheduledTask extends AbstractRunnable {
+
+        final String name;
+        volatile TimeValue interval;
+        volatile ScheduledFuture<?> scheduledFuture;
+
+        public ScheduledTask(String name, TimeValue interval) {
+            this.name = name;
+            this.interval = interval;
+            this.scheduledFuture = null;
+        }
+
+        public void startIfNeeded() {
+            synchronized (mutex) {
+                if (scheduledFuture != null) {
+                    return;
+                }
+                if (shouldRun()) {
+                    logger.debug("scheduling [{}] every [{}]", name, interval);
+                    reschedule();
+                } else {
+                    logger.debug("scheduled [{}] disabled", name);
+                }
+            }
+        }
+
+        /** returns true if this task should be running for this shard and setting */
+        protected boolean shouldRun() {
+            synchronized (mutex) {
+                return interval.millis() > 0 && shouldRunStates().contains(state) && shardRouting.primary();
+            }
+        }
+
+        public void cancelAndStartIfNeeded() {
+            synchronized (mutex) {
+                cancel();
+                startIfNeeded();
+            }
+        }
+
+        public void cancel() {
+            synchronized (mutex) {
+                FutureUtils.cancel(scheduledFuture);
+                scheduledFuture = null;
+            }
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+            if (t instanceof AlreadyClosedException || t instanceof EngineClosedException || state == IndexShardState.CLOSED) {
+                logger.trace("ignoring unexpected error while running [{}], shutting down", t, name);
+            } else {
+                logger.warn("unexpected error while running [{}]", t, name);
+                reschedule();
+            }
+        }
+
+        final EnumSet<IndexShardState> DEFAULT_RUNNING_STATES = EnumSet.of(IndexShardState.RECOVERING,
```

Comment:
Can you double check that `IndexShardState.CREATED` is not needed here too; the `EngineRefresher` use to be eligible to run in that state too?